server/src/test/java/org/elasticsearch/indices/state/CloseWhileRelocatingShardsIT.java

```diff
@@ -88,60 +115,72 @@ public void testCloseWhileRelocatingShards() throws Exception {
+                        currentNodeId = replica.currentNodeId();
+                    }
+                }
 
-                    final String currentNodeId = randomBoolean() ? primary.currentNodeId() : replica.currentNodeId();
-                    assertNotNull(currentNodeId);
-
-                    final Thread thread = new Thread(() -> {
-                        try {
-                            latch.await();
-                        } catch (InterruptedException e) {
-                            throw new AssertionError(e);
+                final DiscoveryNode sourceNode = clusterService.state().nodes().resolveNode(primary.currentNodeId());
+                ((MockTransportService) internalCluster().getInstance(TransportService.class, targetNode))
+                    .addSendBehavior(internalCluster().getInstance(TransportService.class, sourceNode.getName()),
+                        (connection, requestId, action, request, options) -> {
+                            if (PeerRecoverySourceService.Actions.START_RECOVERY.equals(action)) {
+                                logger.debug("blocking recovery of shard {}", ((StartRecoveryRequest) request).shardId());
+                                latch.countDown();
+                                try {
+                                    release.await();
+                                    logger.debug("releasing recovery of shard {}", ((StartRecoveryRequest) request).shardId());
+                                } catch (InterruptedException e) {
+                                    throw new AssertionError(e);
+                                }
+                            }
+                            connection.sendRequest(requestId, action, request, options);
                         }
-                        assertAcked(client().admin().cluster().prepareReroute()
-                            .add(new MoveAllocationCommand(indexToRelocate, shardId, currentNodeId, newNode)));
-                    });
-                    threads.add(thread);
-                    thread.start();
-                }
+                    );
+                commands.add(new MoveAllocationCommand(index, shardId, currentNodeId, targetNode));
             }
 
+            assertAcked(client().admin().cluster().reroute(new ClusterRerouteRequest().commands(commands)).get());
+
             // start index closing threads
+            final List<Thread> threads = new ArrayList<>();
             for (final String indexToClose : indices) {
                 final Thread thread = new Thread(() -> {
                     try {
                         latch.await();
                     } catch (InterruptedException e) {
                         throw new AssertionError(e);
+                    } finally {
+                        release.countDown();
                     }
+                    // Closing is not always acknowledged when shards are relocating: this is the case when the target shard is initializing
```

Comment:
> we need to improve this, can you add it as an item to the meta issue?

Done.

> I wonder if we can fix this by acquiring an operation permit for each batch of operations that we send as part of phase 2 during peer recovery, and then also check whether there's a read-only block under the permit.

I'm not sure to see how it would fix the issue: acquiring an operation permit for a batch of operations does not ensure that all operations have been recovered at the time the verify shard before close action is executed. Or are you thinking of failing the recovery because of the block detected under the permit?

Reply:
yes, I thought about failing the recovery in case where the block suddenly appears during the recovery. This has some other adverse consequences though. Needs more thought